consensus/src/processes/difficulty.rs

```diff
@@ -81,50 +81,54 @@ impl<T: HeaderStoreReader> DifficultyManager<T> {
         // We remove the minimal block because we want the average target for the internal window.
         difficulty_blocks.swap_remove(min_ts_index);
 
+        // We need Uint320 to avoid overflow when summing and multiplying by the window size.
+        // TODO: Try to see if we can use U256 instead, by modifying the algorithm.
         let difficulty_blocks_len = difficulty_blocks.len();
-        let targets_sum: BigInt = difficulty_blocks
+        let targets_sum: Uint320 = difficulty_blocks
             .into_iter()
-            .map(|diff_block| big_from_compact_target(diff_block.bits))
+            .map(|diff_block| Uint320::from(u256_from_compact_target(diff_block.bits)))
             .sum();
         let average_target = targets_sum / (difficulty_blocks_len as u64);
```

Comment:
The overflow of U256 can be avoided by relying on following general idea (pseudo-code, by no means valid rust code):
```
min_diff_block = min(difficulty_blocks)
max_diff_block = max(difficulty_blocks)
offsets_sum = sum(diff_block - min_diff_block)
average_target = min_diff_block + offsets_sum / difficulty_blocks_len
```
This however requires that following condition be true:
`max_diff_block - min_diff_block < 2^256 / difficulty_blocks_len`

Otherwise, an alternate execution path should rely on the current U320 implementation.

Note that the `offsets_sum` approach does floor exactly the same way as the `targets_sum` does.

![AVG floor demonstration](https://user-images.githubusercontent.com/18632023/189611451-473d9445-0931-4289-92d2-21c01e9782c3.jpg)

Reply:
Seems like we will defer this optimization to a following PR